### ISSUE TITLE
Tracking issue: Moving towards Helix captures

### ISSUE THREAD

**Issue by amaanq** [nvim-treesitter/nvim-treesitter#4799]
In an effort to make query collaboration and sharing easier, some of our captures would be better off following upstream's and/or Helix's convention. However, not all of them offer benefits, and some potentially make us worse off. With that being said, the following list below is what I feel would be *good* changes to follow and conform to, discourse welcome!

### [Tree-sitter "standard capture names"](https://github.com/tree-sitter/tree-sitter/blob/3f44b896852eb7daaa6df4fb778c9bb52c70c815/highlight/src/lib.rs#L20-L72):

- [ ] `@parameter` -> `@variable.parameter`
- [ ] `@field` -> `@variable.member`
- [ ] `@namespace` -> `@module`
- [ ] `@float` -> `@number.float`
- [ ] `@symbol` -> `@string.special.symbol`
- [ ] `@string.regex` -> `@string.regexp`
- [ ] `@text.*` -> `@markup.*` (`strong`, `italic`, `link`, `strikethrough`; with exceptions; see below)
- [ ] `@text.title` -> `@markup.heading`
- [ ] `@text.literal` -> `@markup.raw`
- [ ] `@text.reference` -> `@markup.link`
- [ ] `@text.uri` -> `@markup.link.url` (in markup links)
- [ ] `@string.special` -> `@markup.link.label` (non-url links)
- [ ] `@punctuation.special` -> `@markup.list` (markdown lists only; move subitems from `@text.todo`)

### [Helix captures](https://docs.helix-editor.com/master/themes.html#syntax-highlighting):

- [ ] `@method` -> `@function.method`
- [ ] `@method.call` -> `@function.method.call`
- [ ] `@text.{todo,warning,note,danger}` -> `@comment.{error,warning,hint,info,ok}` (Helix uses `@error` etc., but that conflicts with parsing errors)
- [ ] `@text.diff.*` -> `@diff.*` (`plus`, `minus`, `delta`)
- [ ] `@text.uri` -> `@string.special.url` (outside markup)
- [ ] `@preproc` -> `@keyword.directive`
- [ ] `@storageclass` -> `@keyword.storage`

#### Non-changes:

- move control captures to `@keyword.control.*` (too much nesting)
- use other "standard capture names" of questionable value (`@escape`, `@carriage-return`)

### Possible follow-ups

- [ ] distinguish `@comment.line` and `@comment.block`
- [ ] distinguish `@markup.raw.inline` and `@markup.raw.block`
- [ ] distinguish `@type.enum`, `@type.enum.variant`, `@type.interface`
- [ ] deprecate redundant captures (`@property`, ...)

## Roadmap

The following steps will be taken over the next days/weeks:

1. freeze master
2. make changes on main
3. make main the default target for any changes (I'll have to see about moving the CI over)
4. upstream changes (including documentation)
5. make a compat release from master
6. backport query change commits from main to master.

We can discuss whether we want to move the target back to master then (which would be a bit of a mess when rebasing main again, but only once) or keep backporting query and lockfile changes periodically.

**Comment by clason** [comment]
Related question: should we move our `locals.lua` to the [upstream format](https://tree-sitter.github.io/tree-sitter/syntax-highlighting#local-variables)? With the `@local` prefix, this aligns how `indents` and `injections` are named, too.

**Comment by clason** [comment]
Since the `--check` PR has been merged, I think it's safe to consider the captures specified there "offical", and we should start updating at least those.

As this will make merging PR into `main` even harder, I'd like to propose "freezing" `master` at that point, and requiring any changes to be made against `main`. Once we're done, we can backport them to `master` (in bulk) until we declare `main` as default. (At that point, it'll be safe to raise the minimal version to Nvim 0.9 in order to drop the legacy injection format on `master` as well.)

**Comment by clason** [comment]
For the record, the upstream "common core" captures are these: https://github.com/tree-sitter/tree-sitter/blob/3f44b896852eb7daaa6df4fb778c9bb52c70c815/highlight/src/lib.rs#L20-L72

**Comment by clason** [comment]
Actually, some of these would not be breaking, like `@comment` -> `@comment.{line,block}` (or useful fixes like `@string.regex` -> `@string.regexp`) so could already be made on `master`.

**Comment by lucario387** [comment]
When should we start this transition to Helix captures :thinking: 
I think it'll be like 1 or 2 weeks before 1.0 to minimize breakage?

**Comment by clason** [comment]
When the new treesitter version is released, so we can use the built-in checks (and start with the "core upstream captures" this introduced).

**Comment by lucario387** [comment]
`master` ?! Woah that means even 0.9 branches will have to adapt to it?

**Comment by wookayin** [comment]
Feels like we're having another incompatible 0.10.x... I would suggest and prefer (strongly) doing it as a part of v1.0.

**Comment by lucario387** [comment]
> 2. forward the warning to your favorite colorscheme authors

Already forwarded to catppuccin members :D

**Comment by clason** [comment]
Yes. This is driven by the colorscheme updates in core, so needs to be done _before_ we release Nvim 0.10.

The alternative is to do this on `main` **and freeze `master`**. Otherwise keeping `main` rebased will be infeasible.

**Comment by lucario387** [comment]
I'll go with whatever makes Neovim core less of a mess. For that `master` is better since tractions and ~~test in production~~

**Comment by clason** [comment]
I'm happy to make a tag _before_ making these changes and tell people to stay on that tag until the fallout has cleared; I think that would be a good compromise between YOLO breaking and freezing master.

**Comment by ofseed** [comment]
> * [ ]  `@punctuation.special` -> `@markup.list` (move subitems from `@text.todo`)

There's also a `@punctuation.special` in the common core:
https://github.com/tree-sitter/tree-sitter/blob/3f44b896852eb7daaa6df4fb778c9bb52c70c815/highlight/src/lib.rs#L59

Are they the same thing?

**Comment by clason** [comment]
This is just about the "abuse" of `@punctuation.special` for markup lists, not about `@punctuation.special` itself, which we'll keep of course. (As a subcapture, it's implicitly allowed even if not explicitly listed in CONTRIBUTING.md`.)

(If there are things that generally _should_ be `@punctuation.special` instead of something else, feel free to make a suggestion.)

**Comment by ObserverOfTime** [comment]
Can we get a capture for file references so we can stop using `@string.special` (with or without `@text.uri`) for them?

**Comment by clason** [comment]
That would be `@markup.link.label`, according to Helix. I'd lump them all into `@markup.link` for now and leave distinguishing them for a follow-up PR.

**Comment by ObserverOfTime** [comment]
> That would be `@markup.link.label`, according to Helix.

That works for markup files but not file references in code.

**Comment by ObserverOfTime** [comment]
Two more things:

1. We might wanna add `@type.interface` alongside `@type.enum`.

2. Changing `@storageclass` to `@keyword.storage` does not properly address https://github.com/nvim-treesitter/nvim-treesitter/pull/3648#issuecomment-1279923861.

**Comment by clason** [comment]
These issues are orthogonal to this proposal then.

**Comment by ObserverOfTime** [comment]
The first can be a follow-up. The second would be yet another breaking change so it'd be better to address now.

**Comment by clason** [comment]
Separate PR then. Let's keep this one focused. (But again, the goal is to be closer to Helix, not move away from it.)

**Comment by ObserverOfTime** [comment]
OK, here's something for this PR: `@field` and/or `@property` -> `@variable.member`
(The "standard" capture and not Helix's `@variable.other.member`)

**Comment by clason** [comment]
Yeah, `@variable.other` is just weird.

**Comment by lucario387** [comment]
I think distinguishing `@comment.{line,block}` correctly will require neovim/neovim#24738 to be picked up again.

**Comment by clason** [comment]
This is done now; we'll wait for possible cleanup PRs before backporting it.

**Comment by clason** [comment]
Backport is planned for the end of the week. TODO:

* backport `printf` parser
* backport upstream locals
* backport highlight captures
* backport query formatter
* backport test script/CI changes
